### DISCUSSION TITLE
00. PyTorch Fundamentals - Exercises

### DISCUSSION THREAD

**Discussion by hashraf96** [mrdbourke/pytorch-deep-learning#108] [Q&A] [answered]
Hi, I could only verify answers to the questions you (mrdbourke) have posted the answers to.

While going through the exercise for  "00. PyTorch Fundamentals - Exercises", I had some troubles:

```4. Set the random seed to 0 and do exercises 2 & 3 over again.

import torch
seed = 0
torch.manual_seed(seed)
tensor_random = torch.rand(7,7)

torch.manual_seed(seed)
tensor_random_second = torch.rand(1,7)
multiplied = torch.mm(tensor_random, tensor_random_second.T)
multiplied,multiplied.shape
print(f"Tensor A: ",tensor_random )
print(f"\nTensor B: ",tensor_random_second )
print(f"\nTensor B.T: ",tensor_random_second.T )
print(f"\nTensor A * Tensor B.T: ",multiplied )
```
Gives the output:

```Tensor A:  tensor([[0.4963, 0.7682, 0.0885, 0.1320, 0.3074, 0.6341, 0.4901],
        [0.8964, 0.4556, 0.6323, 0.3489, 0.4017, 0.0223, 0.1689],
        [0.2939, 0.5185, 0.6977, 0.8000, 0.1610, 0.2823, 0.6816],
        [0.9152, 0.3971, 0.8742, 0.4194, 0.5529, 0.9527, 0.0362],
        [0.1852, 0.3734, 0.3051, 0.9320, 0.1759, 0.2698, 0.1507],
        [0.0317, 0.2081, 0.9298, 0.7231, 0.7423, 0.5263, 0.2437],
        [0.5846, 0.0332, 0.1387, 0.2422, 0.8155, 0.7932, 0.2783]])

Tensor B:  tensor([[0.4963, 0.7682, 0.0885, 0.1320, 0.3074, 0.6341, 0.4901]])

Tensor B.T:  tensor([[0.4963],
        [0.7682],
        [0.0885],
        [0.1320],
        [0.3074],
        [0.6341],
        [0.4901]])

Tensor A * Tensor B.T:  tensor([[1.5985],
        [1.1173],
        [1.2741],
        [1.6838],
        [0.8279],
        [1.0347],
        [1.2498]])

```
While the author to the question says the answer should be:

```The output should be:

(tensor([[1.8542],
         [1.9611],
         [2.2884],
         [3.0481],
         [1.7067],
         [2.5290],
         [1.7989]]), torch.Size([7, 1]))
```
I don't know what I am doing wrong, can somebody help me with it?

**Comment by aronvandepol** [comment]
From what I can tell, only the values are different, which is likely because of a different manual seed passed onto `torch.rand()`. I believe @mrdbourke always uses 42, the answer to the universe.

**Reply by mrdbourke** [reply]
You're right I generally always use 42 as the random seed but the exercises say to try out various seed values: https://www.learnpytorch.io/00_pytorch_fundamentals/#exercises

**Comment by mrdbourke** [comment, marked as answer]
Hi @hashraf96,

Your code is so close!

The error is only minor. 

The different results come from calling `torch.manual_seed()` more than once in a single code cell.

You generally only need to call it once per code cell.

See the changes I've made to your code: 

```python
import torch
seed = 0

# Only set seed once
torch.manual_seed(seed)

# Create random tensors
tensor_random = torch.rand(7, 7)
tensor_random_second = torch.rand(1, 7)

# Perform matrix multiplication
multiplied = torch.mm(tensor_random, tensor_random_second.T)
multiplied, multiplied.shape

# Print outputs
print(f"Tensor A: ",tensor_random )
print(f"\nTensor B: ",tensor_random_second )
print(f"\nTensor B.T: ",tensor_random_second.T )
print(f"\nTensor A * Tensor B.T: ",multiplied )
```

**Reply by parisppdo** [reply]
Hello!

I encountered the same mistake. I believe the reason why my fellow classmate and me both used `torch.manual_seed()` twice, is the following: In course 36, somewhere around 7:50, you mention that the manual seed only works for one block of code and you use it twice. At first you use it once and it doesn't create "flavoured randomness". You even mention this in block 69 of the textbook.

I searched the whole issue and found out the following:
The `.rand()` method, creates a sequence of numbers that will be included in the tensor. However, this sequence is not fully random. It is the same sequence of numbers each time. However, it appears as random because each time we call it, it will have a random starting point. 
With the `.manual_seed()` method we **set the starting point manually**. As a result, it generates the same sequence each time. After the tensor is created, the "pointer" of the sequence stops at that point.

With this in mind, what happened in the lecture is the following:
You  want to test if the two consecutive tensors you created are the same. You set the manual_seed once. The two tensors that are created are different because their starting point is different. However, when `.manual_seed()` is called again, the pointer is reset to the same starting point and the two tensors are the same. However, I believe that that the phrase "the manual seed only works for a block of code" is wrong. The .manual_seed() method is used to set the starting point of the sequence and it does exactly that in the course as well as both in an IDE or in a notebook (I think that it is mentioned that it works different on each enviroment)

I do not know if I am correct, any feedback is welcome!
